### PR TITLE
Add connections rpc endpoints

### LINKED ISSUE #3293: Add new RPC-style API to create connections
This API is to allow the front end to submit the New Database Connection form, as described in these [specs](https://wiki.mathesar.org/design/specs/new-db-connection-form/).

The endpoint should be `/api/ui/v0/connections/`

Note that this is under the `ui` namespace, not `db`.

Requests should be submitted via POST.

The POST request body, as described via TypeScript, is:

```ts
type SampleDataSchemaIdentifier = 'library_management' | 'movie_collection';

type ConnectionToInternalDatabase = {
  connection_type: 'internal_database';
};
type ConnectionToUserDatabase = {
  connection_type: 'user_database';
  /** The Django ID from the connection object */
  id: number;
};
type Connection = ConnectionToInternalDatabase | ConnectionToUserDatabase;

type CredentialsCopied = {
  strategy: 'copy_from_known_connection';
  connection: ConnectionReference;
};
type CredentialsFromScratch = {
  strategy: 'from_scratch';
  user: string;
  password: string;
  host: string;
  port: string;
};
type CredentialsFromScratch = {
  strategy: 'with_new_user';
  create_user_via: Connection;
  new_user: string;
  new_password: string;
  host: string;
  port: string;
};

type Credentials =
  | CredentialsCopied
  | CredentialsFromScratch
  | CredentialsWithNewUser;

type RequestBody = {
  credentials: Credentials;
  database_name: string;
  create_database: boolean;
  sample_data: SampleDataSchemaIdentifier[];
  nickname: string;
};
```

Examples:

- Case A

    ```json
    {
      "credentials": {
        "strategy": "copy_from_known_connection",
        "connection": {
          "connection_type": "internal_database"
        }
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

- Case B

    ```json
    {
      "credentials": {
        "strategy": "from_scratch",
        "user": "banana",
        "password": "riYwZ1xQ81RW28aU75k",
        "host": "example.com",
        "port": "5432"
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

- Case C

    ```json
    {
      "credentials": {
        "strategy": "with_new_user",
        "create_user_via": {
          "connection_type": "user_database",
          "id": 81
        },
        "user": "banana",
        "password": "riYwZ1xQ81RW28aU75k",
        "host": "example.com",
        "port": "5432"
      },
      "database_name": "mathesar",
      "create_database": true,
      "sample_data": [],
      "nickname": "My new connection"
    }
    ```

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #3293 

This adds 3 endpoints at:

-  `POST api/ui/v0/connections/create_from_known_connection`
-  `POST api/ui/v0/connections/create_from_scratch`
-  `POST api/ui/v0/connections/create_with_new_user`

to allow the creation of connections via the API.

As a bonus, it also adds a function

`msar.create_basic_mathesar_user(username text, password_ text) RETURNS TEXT` to the SQL installation script that can let a DBA easily add a Mathesar user with the proper permissions from the database layer.

## Technical details

All requests are `POST` requests submitting JSON with the following top-level keys

- `credentials` (object, required): Describes at least one set of working credentials (possibly new), and a set of credentials to store (could be the same)
- `database_name` (string, required): The DB-layer name of the database to connect to (or create and then connect to)
- `nickname` (string, required): The user-facing 'friendly' name for the database connection. Must be unique.
- `create_database` (boolean, optional; default false): Whether we should create the DB if it doesn't yet exist
- `sample_data` (array of strings, optional): A list of sample datasets to add to the DB. Possible values are: "library_management" and "movie_collection"

Regarding the `credentials` object: To define the working credentials, one can either give a connection, or one can give user, password, host, and port values.

For a connection, it should be either:

- ```json
  {"connection_type": "internal_database"}
  ```

  Here we use the connection to the internal Django DB to interact with the database server (must be Postgres, of course).

  **or**

- ```json
  {"connection_type": "user_database": "id": 1234}
  ```

  Here we use one of the preexisting DB connections to interact with a database server.

#### Create connection starting from known connection

These should be submitted as `POST` requests to `api/ui/v0/connections/create_from_known_connection/`.

This will create a new connection wired up to a newly-created database on the internal database server. No sample data installed

```json
{
  "credentials": {
    "connection": {
      "connection_type": "internal_database"
    }
  },
  "database_name": "my_mathesar",
  "nickname": "My new connection",
  "create_database": true
}
```

This will create a connection wired up to a newly-created database on whichever server the connection with ID 2 references. Library management schema installed. Possible errors: If the connection with ID 2 isn't working or doesn't have `CREATEDB`, we'll throw an error when we try to install the sample data.

```json
{
  "credentials": {
    "connection": {
      "connection_type": "user_database",
      "id": 2
    }
  },
  "database_name": "anew_mathesar",
  "nickname": "My new connection2",
  "create_database": true,
  "sample_data": ["library_management"]
}
```

#### Create connection starting from scratch

These should be submitted as `POST` requests to `api/ui/v0/connections/create_from_scratch/`.

This will create a new connection using the given credentials. Assuming you're running a normal dev setup, these credentials should work. Otherwise, adjust accordingly. We'll throw an error trying to install sample data if the credentials are invalid.

```json
{
  "credentials": {
    "user": "mathesar",
    "password": "mathesar",
    "host": "mathesar_dev_db",
    "port": "5432"
  },
  "database_name": "mathesar",
  "sample_data": [ "movie_collection" ],
  "nickname": "A Super Connection"
}
```

Here, we use the same setup, but we don't attempt to create the DB or install Mathesar on it, and we don't try to install sample data. In this case, we just store the connection info regardless of whether we can reach the database or not.

```json
{
  "credentials": {
    "user": "mathesar",
    "password": "mathesar",
    "host": "mathesar_dev_db",
    "port": "5432"
  },
  "database_name": "supermathesar",
  "nickname": "A Super Connection 2"
}
```

#### Create connection, along with new user

These should be submitted as `POST` requests to `api/ui/v0/connections/create_with_new_user/`. N.B. You _must_ have a working preexisting connection for these, since you're going to use that to create the user. Either an `internal_database` or `user_database` connection can be used.

```json
{
  "credentials": {
    "create_user_via": {
      "connection_type": "internal_database"
    },
    "user": "banana",
    "password": "bananapass",
  },
  "database_name": "new_user_mathesar",
  "create_database": true,
  "nickname": "My new user's connection"
}
```

The new user will have `CONNECT` and `CREATE` on the given database, as well as `USAGE` privilege on all internal Mathesar schemas.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by mathemancer** [comment]
@Anish9901 This is ready for review, but needs manual testing. Our current testing suite is designed around reusing the database server where the running service has its Django tables, and most of the actions enabled by these endpoints would modify that database without making cleanup guaranteed (for example, if tests are cancelled). Thus, I'm electing to avoid writing API tests (or any tests that use the backing functions) until we have overhauled how the testing suite works.

I'll update the description to make manual testing easier.

**Review by seancolsen — APPROVED** [review]
I haven't tested this exhaustively, but in its most basic form, this is working for me. As far as I'm concerned, we can sort out any remaining deficiencies in subsequent PRs.

**Review by Anish9901 — CHANGES_REQUESTED** [review]
@mathemancer I've a found a few issues,
- Will we be supporting adding databases through the env variable `MATHESAR_DATABASES`? I don't think so. However, if we _are_, this PR removes the error handling for unique violations that occur during server restarts.
- The 1st JSON provided in the PR description for `create_from_scratch` contains the key `create_database: true` which according to the wiki [spec](https://wiki.mathesar.org/design/specs/new-db-connection-form/#case-b-manually-entering-new-connection-credentials) and the implentation in the PR doesn't match up. However the spec and the implementation do contain `sample_data` which requires a database to be present. The JSON fails resulting in a 500 in case the database doesn't exist.
- I tried 2nd JSON for `create_from_scratch` without a database, which resulted in the connection getting created, I then created the database using psql, after which patched the connection with the same JSON but there was no attempt to install mathesar on the database.  
- Executing any of the JSONs twice results in a unique violation with a 500 instead of 400.(due to `db_model.save()` in `_save_and_install`)

File: mathesar/install.py [hunk 1]
```diff
@@ -33,42 +31,19 @@ def main(skip_static_collection=False):
...
+def install_on_db_with_key(database_key, skip_confirm):
+    from mathesar.models.base import Database
+    db_model = Database.create_from_settings_key(database_key)
+    db_model.save()
+    install.install_mathesar(
+        database_name=db_model.db_name,
+        hostname=db_model.host,
+        username=db_model.username,
+        password=db_model.password,
+        port=db_model.port,
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply mathesar/install.py:46]: I don't understand this one. This is using the 'nickname' of databases from the saved DB models.

[the remark this answers is not in the document]
Comment (Anish9901) [reply mathesar/install.py:46]: That's true but notice the last couple of lines of the main func, `install_on_db_with_key` is only called in main and for the dbs present in the `settings.DATABASES` dict.

**Review by Anish9901 — APPROVED** [review]

**Comment by kgodey** [comment]
@Anish9901 @mathemancer Is there a ticket to track the follow up work we discussed at today's meeting, or has that all already been handled?

**Comment by seancolsen** [comment]
@mathemancer you said:

> @seancolsen what do you think about that bit?

Hmmm, it's a little hard for me to tell since I haven't been following this thread between you and Anish so I don't really have enough context to understand the problem. If you'd still like my input would you be able to summarize this into a specific question for me?

**Comment by mathemancer** [comment]
@seancolsen As we discussed and agreed previously, I don't think we should delete or get rid of connections if we fail to connect to the database the connection references. This is because there are a number of reasons that could happen, many of which are _not_ that the DB is deleted. For now, that means throwing an internal server error (500) if we try to reach a database and fail, since that's most reasonable for "couldn't connect between the service and a database". We should get more granular later.

The new question is: What if we can't reach the database when _creating_ a connection? There are multiple scenarios:

- We can reach the server, but can't connect to the DB
  - It doesn't exist
  - It exists, but the user in the connection doesn't have the `CONNECT` privilege on it
- We can't reach the server at all
  - We're copying a connection from a previous connection
    - The previous connection has connected successfully at least once
    - The previous connection has never actually connected

What error(s) should we throw, and what should a user's next steps be?

At the moment, everything is wrapped up in a 500, along with an error message with some more detail about what specifically happened. We still create the connection, for the same reasons we don't delete or remove a connection that isn't working (this is a debatable choice, but makes at least some sense). This is a simplifying choice, but also may compress the situation too much for the user to know how to proceed.

**Comment by mathemancer** [comment]
> @Anish9901 @mathemancer Is there a ticket to track the follow up work we discussed at today's meeting, or has that all already been handled?

I think we need to work out the requirements. I've started an issue here: #3369

**Comment by seancolsen** [comment]
@mathemancer thanks for the summary. These are good questions. I can see advantages and disadvantages to various approaches. I think I would need to talk this through more in order to form opinions on it. The issue #3369 that you opened seems like a good placeholder for that sort of discussion.

**Pull request merged**